Add functionality to compute membership inference risk for each individual point

lwsong opened the pull request:
Following our recent paper https://arxiv.org/abs/2003.10595, I implement the code to compute the privacy risk score for each individual sample, which represents its likelihood of being a member. The main function is defined as "_compute_privacy_risk_score" in "membership_inference_attack.py". The function will compute risk scores for all training and test points, which are passed to the "SingleRiskScoreResult" class in "data_structures.py". I also add "codelab_privacy_risk_score.ipynb" to demonstrate how to run the code. Test cases for privacy risk scores are also added.

CdavM reviewed (approved):
Looks good overall. I will probably need to make some formatting changes for consistency (like the printing format). Hopefully this will be merged by the end of the week.

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -529,6 +529,39 @@ def __str__(self):
+  
+  test_risk_scores: np.ndarray
+  
+  def attack_with_varied_thresholds(self, threshold_list):
```
Can you add a comment for this function?

lwsong replied:
Just added the comment!

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -529,6 +529,39 @@ def __str__(self):
+@dataclass
+class SingleRiskScoreResult:
+  """Results from computing privacy risk scores.
+  this part is quite preliminary: it shows how to leverage privacy risk score to perform attacks with thresholding on risk score
```
What does it mean for this part to be preliminary? Is it going to be updated in the future?

lwsong replied:
By saying preliminary, I mean that we may leverage privacy risk scores for more analysis in the future. 
Currently, I follow our paper to set a threshold on the privacy risk score and compute precision and recall values.
With full training and test privacy risk scores, we can also return sample indices that have high privacy risk scores or implement other functionalities based on what we want.
I change the comment a bit in the updated code.

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -529,6 +529,39 @@ def __str__(self):
+          recall_list.append(true_positive_normalized)
+    return meaningful_threshold_list, precision_list, recall_list
+  
+  def print_results(self, threshold_list=[1,0.9,0.8,0.7,0.6,0.5]):
```
Why is the default value this threshold list? Can you add a comment and a test for this function?

lwsong replied:
I added a comment to it. Will further edit this function later

lwsong replied:
Just rewrote this function!

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -529,6 +529,39 @@ def __str__(self):
+  def print_results(self, threshold_list=[1,0.9,0.8,0.7,0.6,0.5]):
+    meaningful_threshold_list, precision_list, recall_list = self.attack_with_varied_thresholds(threshold_list)
+    for i in range(len(meaningful_threshold_list)):
+      print(f"with {meaningful_threshold_list[i]} as the threshold on privacy risk score, the precision-recall pair is {(precision_list[i], recall_list[i])}")
```
Instead of calling print every line, we can just create a summary string and return that. See https://github.com/tensorflow/privacy/blob/b208d9deece83bc251d5a7ca6b54612f06885eae/tensorflow_privacy/privacy/membership_inference_attack/data_structures.py#L510 for an example

lwsong replied:
Fixed it!

CdavM reviewed (commented):
Hi @lwsong ,I took another look and left some minor comments. Also, can you add the required types in function signatures (and change regular lists to np.ndarrays)?

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/codelabs/codelab_privacy_risk_score.ipynb:
```diff
@@ -0,0 +1,816 @@
+    "build_target": "//learning/deepmind/public/tools/ml_python:ml_notebook",
+    "kind": "private"
+   },
+   "name": "Membership inference codelab",
```
Should we rename this to make it clear that this is about the privacy risk score?

Also, can we rename the filename to privacy_risk_score_codelab to make it consistent with other files?

lwsong replied:
Fixed these issues!

CdavM commented on tensorflow_privacy/privacy/membership_inference_attack/codelabs/codelab_privacy_risk_score.ipynb:
```diff
@@ -0,0 +1,816 @@
+    "                            by_classification_correctness = True\n",
+    "                            )\n",
+    "input_slice_specs = get_single_slice_specs(slicing_spec, 10)\n",
+    "for single_slice_spec in input_slice_specs:\n",
```
So is the individual level privacy risk score meant for slices or individual samples? I was naively expecting something like: out of all the samples in the training set, these are especially likely to be memorized. 

But this example makes me think that we're still doing this on groups of samples (like slices)

lwsong replied:
We actually compute a privacy risk score for each individual sample: in data_structures.py, the SingleRiskScoreResult data class has input arguments of train_risk_scores and test_risk_scores, each element of them means a sample's privacy risk score.

I call the slice function mainly to follow what we do in detailed attacks, where the dataset is also divided into slices.
If you want, we can find time to chat more about this.

lwsong commented:
Hi @CdavM , I just fixed all these issues you mentioned. Please take a look, thanks!

CdavM commented:
Hi @lwsong !

Thanks for making these changes. This looks great! I can't merge the PR, the button is grayed out and there's a warning saying "This branch has conflicts that must be resolved". Can you resolve the conflicts?

Also, a meeting would be great. I worry that finding a good time with the holiday season will be hard, but I'll set something up in early January. The meeting isn't blocking this PR anyway.

lwsong commented:
Hi @CdavM , just resolved the conflicts. Let me know if there is an issue.
Sure, let's meet in early January.

dvadym commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  
+  min_log_value = np.amin(np.concatenate((train_log_values, test_log_values)))
+  max_log_value = np.amax(np.concatenate((train_log_values, test_log_values)))
+  bins_hist = np.linspace(min_log_value, max_log_value, num_bins+1)
```
It looks replacement 
np.linspace -> np.logspace, would make log histograms on train_values/test_values (i.e. w/o logs).
Is it correct replacement?
If yes, could you please update the code? (it's always better to have simpler code for understanding and maintenance)

lwsong replied:
Changed the code to np.logspace!

dvadym commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  
+  train_hist, _ = np.histogram(train_log_values, bins=bins_hist)
+  train_hist = train_hist/(len(train_log_values)+0.0)
+  train_hist_indices = np.fmin(np.digitize(train_log_values, bins=bins_hist),num_bins)-1
```
why -1? doesn't np.digitize return 0-based indices?

lwsong replied:
bins_hist has num_bins+1 elements, and np.digitize returns values from 1 to num_bins+1 (https://stackoverflow.com/questions/40880624/binning-in-numpy)

dvadym commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -444,6 +444,69 @@ def __str__(self):
+    precision_list = []
+    recall_list = []
+    meaningful_threshold_list = []
+    for threshold in threshold_list:
```
it looks that some of these computations possible to do with sklearn and already implemented classes (doing with the standard APIs is good for understanding and maintenance), namely
1..Compute RocCurve in privacy_risk_score_analysis() as in 
https://github.com/tensorflow/privacy/blob/b208d9deece83bc251d5a7ca6b54612f06885eae/tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py#L74
2.ROCCurve has tpr, fpr, thresholds (ordered). Then here is to get the index of each element of threshold_list in roc_curve.threshould and to get tpr/fpr from roc_curve

Would it work? WDYT?

BTW, does it make sense to return ROC curve and ROC params (like AUC) and to plot charts of RoCCurve

lwsong replied:
I updated the code with metrics.roc_curve function.
I am not sure whether we should return ROC curve and ROC params here, it just makes the privacy risk score analysis very similar to the previous attacks. So I did not change the function further.

Now we compute the privacy risk score for every training and test point, one way to show the result is that we choose a threshold value on privacy risk score (e.g., 0.8), if a training sample's privacy risk score is larger than the threshold, we think it is highly privacy-risky, then we can tell the customers how many risky points we have and further the indices of those risky points.
I will suggest a meeting after the holiday season to discuss more on how we should present the privacy risk score results.

dvadym commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -444,6 +444,69 @@ def __str__(self):
+          recall_list.append(true_positive_normalized)
+    return np.array(meaningful_threshold_list), np.array(precision_list), np.array(recall_list)
+  
+  def collect_results(self, threshold_list=np.array([1,0.9,0.8,0.7,0.6,0.5])):
```
Using default arguments with mutable types is not allowed by Google Python Styleguide
https://google.github.io/styleguide/pyguide.html#212-default-argument-values

Please remove using default arguments here.

The reason is that it's error-prone (more details in style guide)

lwsong replied:
removed the default arguments

dvadym reviewed (changes requested):
Thanks for contributing, it looks great! 

I left comments. Sorry maybe there are more of them, it's very common in software development to have more comment on code review. Of course I might have missed something in comments or wrote something incorrect, that please respond (just for the context, usually comments on code review are kind of the discussions).

sushkoy commented on tensorflow_privacy/privacy/membership_inference_attack/codelabs/privacy_risk_score_codelab.ipynb:
```diff
@@ -0,0 +1,804 @@
+   "source": [
+    "<table class=\"tfo-notebook-buttons\" align=\"left\">\n",
+    "  <td>\n",
+    "    <a target=\"_blank\" href=\"https://colab.research.google.com/github/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/membership_inference_attack/codelabs/codelab.ipynb\"><img src=\"https://www.tensorflow.org/images/colab_logo_32px.png\" />Run in Google Colab</a>\n",
```
Minor: This link and the link below seem to point to an incorrect colab.

sushkoy commented on tensorflow_privacy/privacy/membership_inference_attack/codelabs/privacy_risk_score_codelab.ipynb:
```diff
@@ -0,0 +1,804 @@
+    "\n",
+    "This part shows how to use the privacy risk score.\n",
+    "\n",
+    "For each data slice, we compute privacy risk scores for both training and test data. We then set a threshold on risk scores (an input is inferred as a member if and only if its risk score is higher than the threshold) and compute the attack precision and recall values"
```
Would that be possible to add some interpretation of the obtained results?

Some questions that we could highlight:
 - How does the precision / recall figures compare to the membership inference attack? As you have the results for MIA right above, it might be helpful to compare the two methods. 
- Are there any samples that have high risk scores? Is there anything special about those samples?
 - What is the distribution of privacy risk scores? We could probably plot a simple histogram.

lwsong replied:
For the precision-recall figures, it will be pretty similar to the threshold attacks based on prediction loss or entropy since the privacy risk score is computed based on the distributions of prediction loss or entropy over training and test data.
The importance of privacy risk score analysis is that we actually compute a risk value for each sample such that we can know which samples have high risks. The precision-recall metric is just one way to present the results.

Indeed, some samples have high risk scores. In the codelab running example, class 3 and class 4 have certain training samples with the risk score of 1. It is a very interesting future direction to explore why those samples have high risks.

I agree that we can plot the histogram of the privacy risk scores.
I will suggest a meeting after the holiday season to thoroughly discuss what is the best way to present the privacy risk score results.

sushkoy replied:
Setting up a meeting sounds good, thank you. Some comments (not blocking for this PR, we can address those after the meeting):

`it will be pretty similar to the threshold attacks based on prediction loss or entropy since the privacy risk score is computed based on the distributions of prediction loss or entropy over training and test data.`

This sounds reasonable. Can we compare these numbers in the codelab, as a sanity check?

`It is a very interesting future direction to explore why those samples have high risks.`

As a low hanging fruit, could we plot a few pictures with low and high risk? That might already be very insightful.
We're looking for some way to show how this metrics work in practice.

`importance of privacy risk score analysis is that we actually compute a risk value for each sample such that we can know which samples have high risks`

This makes me realize that some of the trained attacks, and the threshold attack also give some kind of score, which we threshold to classify into training / non-training examples. For the threshold attack, this score is simply the loss. For the neural network classifiers, this is the output of the softmax layer.

Conceptually your score a very similar, except that you provide a different empirical measure.

We discussed this internally, and we think it might make sense to consolidate this with the other membership attacks instead of having this as a separate codepath. This might make the implementation cleaner and simpler.

We're happy to do this refactoring ourselves, just wanted to let you know about this.

lwsong replied:
I added the AUC and advantage values into the new codelab file. As expected, the results are similar to threshold attacks.

I also plotted several figures with high and low risks for each class label in the codelab file, please take a look.

I agree that neural network attack classifiers or threshold attacks can also give some kind of scores.
The advantage of our proposed privacy risk score metric is that it closely captures the real likelihood of being in the training set. For example, if we count all training and test samples with the privacy risk score around (let's say) 0.8, then among those samples, it is indeed around 80% of them are from the training set. While neural network classifier output usually does not well capture the real likelihood of being a member. You can check more in our USENIX paper (Figure 3 and Figure 11 in https://arxiv.org/pdf/2003.10595.pdf)

Sure, I can see the benefits of consolidating the privacy risk score part with other attacks.

sushkoy reviewed (commented):
Thanks a lot! Added a small question on the colab. It would be helpful to add interpretation/discussion of results.

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+                               test_risk_scores=test_risk_scores)
+
+
+def privacy_risk_score_analysis(attack_input: AttackInputData,
```
already renamed it

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
 
+def _compute_privacy_risk_score(attack_input: AttackInputData,
+                                num_bins: int = 15) -> SingleRiskScoreResult:
+  """compute each individual point's likelihood of being a member (https://arxiv.org/abs/2003.10595)
```
Fixed it

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  """compute each individual point's likelihood of being a member (https://arxiv.org/abs/2003.10595)
+  Args:
+    attack_input: input data for compute privacy risk scores
+    num_bins: the number of bins used to compute the training/test histogram; we set the default as 15
```
sure, deleted it in the updated code

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
 
+def _compute_privacy_risk_score(attack_input: AttackInputData,
+                                num_bins: int = 15) -> SingleRiskScoreResult:
+  """compute each individual point's likelihood of being a member (https://arxiv.org/abs/2003.10595)
```
Added more explanation in the comments

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  if attack_input.loss_train is not None and attack_input.loss_test is not None:
+    train_values, test_values = attack_input.loss_train, attack_input.loss_test
+  elif attack_input.entropy_train is not None and attack_input.entropy_test is not None:
+    train_values, test_values = attack_input.entropy_train, attack_input.entropy_test
```
fixed it

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  elif attack_input.entropy_train is not None and attack_input.entropy_test is not None:
+    train_values, test_values = attack_input.entropy_train, attack_input.entropy_test
+  else:
+    train_values, test_values = attack_input.get_loss_train(), attack_input.get_loss_test()
```
fixed it

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/membership_inference_attack.py:
```diff
@@ -172,6 +174,85 @@ def run_attacks(attack_input: AttackInputData,
+  train_log_values = np.log(np.maximum(train_values, small_value))
+  test_log_values = np.log(np.maximum(test_values, small_value))
+  
+  min_log_value = np.amin(np.concatenate((train_log_values, test_log_values)))
```
updated the code!

lwsong commented on tensorflow_privacy/privacy/membership_inference_attack/data_structures.py:
```diff
@@ -444,6 +444,69 @@ def __str__(self):
+    summary = []
+    for single_result in self.risk_score_results:
+      single_summary = single_result.collect_results()
+      for line in single_summary:
```
already updated the code!

lwsong commented:
Hi @dvadym @sushkoy , I updated the code following your comments. For the codelab file, I feel it is better to have a meeting to thoroughly discuss what is the best way to present the privacy risk score results before I further change the code. So I just leave the file as it is. Let me know if you have more comments or questions, thanks!

dvadym reviewed (commented):
Thanks for addressing comments!

We've discussed internally about the metric name. Risk is very generic term and for library users (which might be unfamiliar with details) it might create a wrong impression, eg. when the training data are public there is no risk in exposure any information about the train data. It's better to represent in the name more precisely what it means.
Would it be possible to have some name that represents it? Eg. train_probability (just for example). WDYT?
It's fine to have in comments references to privacy score, since it's how it's named in the paper.

lwsong commented:
Hi @dvadym @sushkoy , I have added more results (AUC and advantage values, figure plots) in the codelab file.
For the name, I have changed "risk score" to "membership_probability" in all related code, including the codelab. Let me know if there are other questions or comments, thanks.

dvadym reviewed (approved):
Thanks!
